### DISCUSSION TITLE
单独编译 `fduthesis` 用户手册时，代码摘录环境缩进问题

### DISCUSSION THREAD

**Discussion by xkwxdyy** [stone-zeng/fduthesis#254] [General]
我 clone 最新的 `fduthesis` 仓库（最新一次提交为 a1bf5281f2289a28cb6cb5bee1f0c79ab0981801 ），并且编译了 `source/fduthesis.dtx` 得到了用户手册，其中代码摘录的效果缩进正常。、

![image](https://user-images.githubusercontent.com/87134063/164957134-9ac6ebbe-7274-4541-8d18-55e73837f2fd.png)

但是我将 <https://github.com/stone-zeng/fduthesis/blob/main/source/fduthesis.dtx#L251-L2632> 的代码摘出来放进了 `tex` 文件中，并将 `^^A!` 等内容去掉之后进行编译，发现代码摘录少了一个缩进，这是什么原因呢？

![image](https://user-images.githubusercontent.com/87134063/164957151-f33ab54c-a90d-4bab-8fb8-1e8e051c2ac0.png)

这是我摘录后处理好的文件：
[demo.zip](https://github.com/stone-zeng/fduthesis/files/8549044/demo.zip)

系统：mac，发行版：texlive2022

**Comment by syvshc** [comment]
原因在这里:

```latex
\lstdefinestyle{style@base}
  {
    basewidth       = 0.5 em,
    gobble          = 3,
    lineskip        = 2 pt,
    frame           = l,
    framerule       = 1 pt,
    framesep        = 0 pt,
    escapeinside    = {(*}{*)},
    basicstyle      = \small\ttfamily\color{MaterialGrey900},
    keywordstyle    = \bfseries\color{MaterialIndigo},
    commentstyle    = \itshape\color{MaterialGrey600},
    stringstyle     = \color{MaterialRed},
    backgroundcolor = \color{MaterialGrey50}
  }
```

其中的 `gobble=3` 会吃掉环境内部的前三个字符, 而 `shellexample` 环境用了 `style@base` 这个样式, 并且

```latex
\begin{shellexample}[morekeywords={tlmgr,install}]
  tlmgr install fduthesis
\end{shellexample}
```

中 `tlmgr` 前面只有两个字符, `gobble=3` 的时候自然 `t` 也被吃掉了, 变成了 `lmgr`, 如果需要改全局样式的话, 可以直接在 `style@base` 处进行修改, 或者使用

```latex
\begin{shellexample}[gobble=2, morekeywords={tlmgr,install}]
  tlmgr install fduthesis
\end{shellexample}
```

整个文档后面代码环境也是同理.

**Reply by xkwxdyy** [reply]
那为什么 dtx 文件里出来是正确的呢
里面的内容相对于环境的缩进应该是一样的

**Reply by xkwxdyy** [reply]
我指的是源代码在写的时候的缩进

**Reply by syvshc** [reply]
源代码里除去 `%^^A!` 后还有三个空格

```latex
%^^A! \begin{shellexample}[morekeywords={tlmgr,install}]
%^^A!   tlmgr install fduthesis
%^^A! \end{shellexample}
```

在 VSCode 取消注释的时候会自动取消一个空格, 对比

```latex
\begin{shellexample}[morekeywords={tlmgr,install}]
  tlmgr install fduthesis
\end{shellexample}
 \begin{shellexample}[morekeywords={tlmgr,install}]
   tlmgr install fduthesis
 \end{shellexample}
```
